### PR TITLE
beforeAll decorator

### PR DESCRIPTION
**Pull request by SergejKasper**
This pull request:
- adds the beforeAll decorator
- makes sure beforeAlls of parent-classes are considered with and without @suite
- makes sure beforeAlls of parent-classes are considered in child suites before the child beforeAlls 
- makes sure tests of parent-classes are not considered in child suites and child tests are not considered in parent suites

### CODE REVIEW

File: src/__tests__/test-before-all.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import { getBeforeAlls, beforeAll as beforeAllDecorator } from "../before-all"
+import { configure, It, BeforeAll } from "../configure"
+
+describe("`beforeAll`", () => {
+    let mockIt: jest.Mock<It>
+    let mockItOnly: jest.Mock<It>
+    let mockBeforeAll: jest.Mock<BeforeAll>
+    beforeAll(() => {
+        mockIt = jest.fn()
+        mockItOnly = jest.fn()
```
Comment (Prior99) [inline src/__tests__/test-before-all.ts]: Is this needed?

Comment (Prior99) [reply src/__tests__/test-before-all.ts]: I think `itOnly` doesn't need to be configured anywhere in this file.

File: src/__tests__/test-configure.ts [hunk 2]
```diff
@@ -2,6 +2,16 @@ import { configure, configuration, initializeConfiguration } from "../configure"
 
 test("initializing the configuration in a browser", () => {
     initializeConfiguration()
+    expect(configuration.beforeEach).toBe(beforeEach)
```
Comment (Prior99) [inline src/__tests__/test-configure.ts:5]: Did you mean to duplicate this test?

File: src/__tests__/test-suite.ts [hunk 3]
```diff
@@ -1,81 +1,233 @@
...
-        mockItOnly = jest.fn()
-        mockDescribe = jest.fn()
-        mockDescribeOnly = jest.fn()
-        configure({
-            it: mockIt,
-            itOnly: mockItOnly,
-            describe: mockDescribe,
-            describeOnly: mockDescribeOnly,
-        })
+  beforeEach(() => {
```
Comment (Prior99) [inline src/__tests__/test-suite.ts]: You changed the indentation in the whole file. Please revert these changes.

File: src/__tests__/test-suite.ts [hunk 4]
```diff
@@ -1,81 +1,233 @@
...
+      public someTest() { mockTestImpl(this) }
+    }
+    expect(mockDescribe).toHaveBeenCalledTimes(1)
+    mockDescribe.mock.calls[0][1]()
+    expect(mockIt).toHaveBeenCalledTimes(1)
+    mockIt.mock.calls[0][1]()
+    expect(mockTestImpl).toHaveBeenCalledTimes(1)
+    expect(mockTestImpl.mock.calls[0][0].constructor).toBe(A)
+  })
+  test("beforeAll-functions are called in proper order before the tests and sets test values reliabily", () => {
```
Comment (Prior99) [inline src/__tests__/test-suite.ts]: Typo: `reliabily` => `reliably`

File: src/__tests__/test-suite.ts [hunk 5]
```diff
@@ -1,81 +1,233 @@
...
+      @beforeAllDecorator
+      public setInitialConditionsOnInstance2() {
+        this.initialConditions += " value"
+      }
+      @testDecorator
+      public runTestWithPresetConditions() {
+        mockTestImpl(this.initialConditions)
+      }
+    }
+    expect(mockDescribe).toHaveBeenCalledTimes(1)
```
Comment (Prior99) [inline src/__tests__/test-suite.ts]: The asserts

```
expect(mockDescribe).toHaveBeenCalledTimes(1)
mockDescribe.mock.calls[0][1]()
expect(mockIt).toHaveBeenCalledTimes(1)
mockIt.mock.calls[0][1]()
expect(mockTestImpl).toHaveBeenCalledTimes(1)
expect(mockTestImpl.mock.calls[0][0]).toBe("some initial value")
```

are not related to this test and already tested in another test above. Please remove them for the sake of readability.

File: src/__tests__/test-suite.ts [hunk 6]
```diff
@@ -1,81 +1,233 @@
...
+      @beforeAllDecorator
+      public setInitialConditionsOnInstance() {
+        this.initialConditions += " inherited"
+      }
+      @testDecorator
+      public runTestWithPresetConditions() {
+        mockTestImpl(this.initialConditions)
+      }
+    }
+    expect(mockDescribe).toHaveBeenCalledTimes(2)
```
Comment (Prior99) [inline src/__tests__/test-suite.ts]: See above. Please remove the unrelated asserts.

File: src/__tests__/test-suite.ts [hunk 7]
```diff
@@ -1,81 +1,233 @@
...
+    mockDescribe.mock.calls[0][1]()
+    mockDescribe.mock.calls[1][1]()
+    expect(mockBeforeAll).toHaveBeenCalledTimes(3)
```
Comment (Prior99) [inline src/__tests__/test-suite.ts]: Why is it 3 times?

File: src/__tests__/test-suite.ts [hunk 8]
```diff
@@ -1,81 +1,233 @@
...
+        params: [
+          { a: 1, b: 2 },
+        ],
+      })
+      protected testSomethingInTheSuperClass(...args: any[]) { return }
+    }
+    @suite
+    class A extends SuperA {
+      /**
+       * This test should only be executed in the child
```
Comment (Prior99) [inline src/__tests__/test-suite.ts]: Please end all sentences with a `.`

File: src/__tests__/test-suite.ts [hunk 9]
```diff
@@ -1,81 +1,233 @@
...
+          { a: 1, b: 2 },
+        ],
+      })
+      protected testSomethingInTheExtendingClass(...args: any[]) { return }
+    }
+    expect(mockDescribe).toHaveBeenCalledTimes(2)
+    mockDescribe.mock.calls[0][1]()
+    mockDescribe.mock.calls[1][1]()
+    // There should be two calls because the parent should not execute the childs test
+    // and the child should not execute the parent tests
```
Comment (Prior99) [inline src/__tests__/test-suite.ts]: Please end all sentences with a `.`

File: src/__tests__/test-suite.ts [hunk 10]
```diff
@@ -1,81 +1,233 @@
...
+    expect(mockIt).toHaveBeenCalledTimes(2)
+    expect(mockIt.mock.calls).toMatchSnapshot()
+  })
```
Comment (Prior99) [inline src/__tests__/test-suite.ts]: Please add an empty line between tests.

File: src/__tests__/test-suite.ts [hunk 11]
```diff
@@ -1,81 +1,233 @@
...
+    mockBeforeAll.mock.calls[2][0]()
+    expect(mockIt).toHaveBeenCalledTimes(3)
+    mockIt.mock.calls[0][1](new A())
+    mockIt.mock.calls[1][1](new A())
+    mockIt.mock.calls[2][1](new A())
+    expect(mockTestImpl).toHaveBeenCalledTimes(1)
+    expect(mockTestImpl.mock.calls[0][0]).toBe("initial value inherited")
+    expect(mockTestImplSuper).toHaveBeenCalledTimes(2)
+    expect(mockTestImplSuper.mock.calls[0][0]).toBe("initial value")
+  })
```
Comment (Prior99) [inline src/__tests__/test-suite.ts]: Please add an empty line between tests.

File: src/__tests__/test-suite.ts [hunk 12]
```diff
@@ -1,81 +1,233 @@
...
+    expect(mockDescribe).toHaveBeenCalledTimes(1)
+    mockDescribe.mock.calls[0][1]()
+    expect(mockBeforeAll).toHaveBeenCalledTimes(2)
+    mockBeforeAll.mock.calls[0][0]()
+    mockBeforeAll.mock.calls[1][0](new A())
+    expect(mockIt).toHaveBeenCalledTimes(1)
+    mockIt.mock.calls[0][1]()
+    expect(mockTestImpl).toHaveBeenCalledTimes(1)
+    expect(mockTestImpl.mock.calls[0][0]).toBe("some initial value")
+  })
```
Comment (Prior99) [inline src/__tests__/test-suite.ts]: Please add an empty line between tests.

File: src/__tests__/test-suite.ts [hunk 13]
```diff
@@ -1,81 +1,233 @@
...
+      @testDecorator
+      public someTest() { mockTestImpl(this) }
+    }
+    expect(mockDescribe).toHaveBeenCalledTimes(1)
+    mockDescribe.mock.calls[0][1]()
+    expect(mockIt).toHaveBeenCalledTimes(1)
+    mockIt.mock.calls[0][1]()
+    expect(mockTestImpl).toHaveBeenCalledTimes(1)
+    expect(mockTestImpl.mock.calls[0][0].constructor).toBe(A)
+  })
```
Comment (Prior99) [inline src/__tests__/test-suite.ts]: Please add an empty line between tests.

File: src/__tests__/test-suite.ts (hunk 9 above)
Comment (Prior99) [reply src/__tests__/test-suite.ts]: You also did this below. I'd be glad if you'd have a look and fix it everywhere.

File: src/__tests__/test-test.ts [hunk 14]
```diff
@@ -76,7 +76,7 @@ describe("`test`", () => {
         }
         const tests = getTests(A)
         expect(tests.length).toBe(3)
-        tests.forEach(storedTest => storedTest(undefined))
+        tests.forEach(storedTest => storedTest(new A()))
```
Comment (Prior99) [inline src/__tests__/test-test.ts:79]: Good catch!

File: src/before-all.ts [hunk 15]
```diff
@@ -0,0 +1,50 @@
+import "reflect-metadata"
+import { TestOptionsInput, parseTestOptions, isTestOptionsInput } from "./options"
+import { configuration } from "./configure"
+
+export type BeforeAllFunction = (instance: any) => void
+
+export function getBeforeAlls(target: Object): BeforeAllFunction[] {
+    const beforeAlls = Reflect.getMetadata("test-decorator:beforeAll", target)
```
Comment (Prior99) [inline src/before-all.ts]: in `test.ts` the name of the reflection key is plural, please name them consistently and either change `test-decorator:tests` to `test-decorator:test` or `test-decorator:beforeAll` to `test-decorator:beforeAlls`. Now that I am thinking about this, maybe we should rename the prefix `test-decorator` to `test-decorators` as it would then match the name of the package.

File: src/before-all.ts [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+    if (typeof beforeAlls !== "undefined") { return beforeAlls }
+    const newBeforAllContainer: BeforeAllFunction[] = []
```
Comment (Prior99) [inline src/before-all.ts]: Typo: `newBeforAllContainer` => `newBeforeAllContainer`

File: src/before-all.ts [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+    Reflect.defineMetadata("test-decorator:beforeAll", newBeforAllContainer, target)
+    return newBeforAllContainer
+}
+
+export type TestDecorator =
```
Comment (Prior99) [inline src/before-all.ts]: I think this should be called `BeforeAllDecorator` as it is the the signature of the beforeAll decorator and not the test decorator. As it has the same signature as the `TestDecorator` and perhaps others will also have, we should move this into a file `types.ts` or similar and call it `SuiteMethodDecorator` or similar.

File: src/before-all.ts [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+    // Load the `beforeAll` function from the configuration which is stored singleton.
+    // It might have been configured to overwrite the mocha ones.
+    const { beforeAll: beforeAllFn } = configuration
+    if (typeof beforeAllFn !== "function") {
+        throw new Error("'beforeAll' not found. Did you call 'configure'?")
+    }
+    const decorator: TestDecorator = (target, property, descriptor) => {
+        const beforeAlls = getBeforeAlls(target.constructor)
+        beforeAlls.push(instance => {
+            // prevents super-class from executing the beforeAll-functions of extending classes
```
Comment (Prior99) [inline src/before-all.ts]: Please use full english sentences. Sorry for being so strict.

```
// prevents super-class from executing the beforeAll-functions of extending classes
```
=>
```
// Prevents the super-class from executing the `beforeAll`-functions of extending classes.
```

File: src/configure.ts [hunk 19]
```diff
@@ -8,6 +13,14 @@ export type DescribeFunction = () => Promise<void> | void
  * Options provided to `configure` for setting up the test framework.
  */
 export interface ConfigureOptions {
+    /**
+     * A function called before all tests
```
Comment (Prior99) [inline src/configure.ts]: Please add a dot at the end of the sentence. This doc is shown to IDE users when autocompleting.

File: src/configure.ts [hunk 20]
```diff
@@ -8,6 +13,14 @@ export type DescribeFunction = () => Promise<void> | void
...
+     */
+    beforeAll?: BeforeAll,
+    /**
+     * A function called for each test
```
Comment (Prior99) [inline src/configure.ts]: Better: `A function called before each test.`

File: src/configure.ts (hunk 19 above)
Comment (Prior99) [reply src/configure.ts]: Better: `A method called once per suite before all tests.`

File: src/test.ts [hunk 21]
```diff
@@ -4,11 +4,11 @@ import { configuration } from "./configure"
 
 export type TestFunction = (instance: any) => void
 
-export function getTests(target: Object): TestFunction[] {
-    const tests = Reflect.getMetadata("test-decorator:tests", target)
+export function getTests(target: any): TestFunction[] {
+    const tests = Reflect.getMetadata(`test-decorator:tests`, target)
```
Comment (Prior99) [inline src/test.ts]: why did you change to template string literals?

File: src/test.ts [hunk 22]
```diff
@@ -56,13 +56,18 @@ export function test<Params>(
         // If `params` was not specified, call `it` once with no parameters.
         if (!params) {
             tests.push(instance => {
+                // prevents super-class from executing the tests of extending classes
```
Comment (Prior99) [inline src/test.ts]: See comment in `before-all.ts`

File: src/test.ts [hunk 23]
```diff
@@ -56,13 +56,18 @@ export function test<Params>(
...
+                if (!(instance instanceof target.constructor)) { return }
                 itFunction(name(), async (...args: any[]) => await descriptor.value.apply(instance, args))
             })
             return descriptor
         }
         // Call `it` once for all parameters configured in `params`.
         tests.push(...params.map(param => {
             return (instance: any) => {
+                // prevents super-class from executing the tests of a child class
```
Comment (Prior99) [inline src/test.ts]:
```
Prevents the super-class from executing the tests of a child class,
while also preventing the child class from executing the tests of the super-class.
```

**Review by Prior99 — CHANGES_REQUESTED** [review]
Hey. Great work, also about the catches you with me calling the decorators the wrong way in the tests. Good catch about the inheritance also. I am sorry for my strictness regarding wording and comments, but I like to keep the code of libraries I maintain exceptionally clean. Please fix my comments. I am open about any discussions and questions that might come up.

File: src/__tests__/test-before-all.ts (hunk 1 above)
Comment (SergejKasper) [reply src/__tests__/test-before-all.ts]: good catch!

File: src/__tests__/test-configure.ts (hunk 2 above)
Comment (SergejKasper) [reply src/__tests__/test-configure.ts:5]: Thanks!!

File: src/__tests__/test-suite.ts (hunk 6 above)
Comment (SergejKasper) [reply src/__tests__/test-suite.ts]: As discussed we leave this in here to make sure that the methods are not called more often than expected when classes are extended

File: src/before-all.ts (hunk 17 above)
Comment (SergejKasper) [reply src/before-all.ts]: thx! nice catch

File: src/test.ts (hunk 21 above)
Comment (SergejKasper) [reply src/test.ts]: I originally tried to provide inheritance collisions the tests based on the constructor name of the target

File: src/__tests__/test-configure.ts [hunk 24]
```diff
@@ -2,16 +2,7 @@ import { configure, configuration, initializeConfiguration } from "../configure"
...
-    expect(configuration.itOnly).toBe(it.only)
-    expect(configuration.describe).toBe(describe)
-    expect(configuration.describeOnly).toBe(describe.only)
-})
-
-test("initializing the configuration in a browser", () => {
-    initializeConfiguration()
-    expect(configuration.beforeEach).toBe(beforeEach)
+    expect(configuration.beforeAll).toBe(beforeAll)
     expect(configuration.it).toBe(it)
```
Comment (Prior99) [inline src/__tests__/test-configure.ts]: isn't a `expect(configuration.beforeEach).toBe(beforeEach)` missing?

File: src/__tests__/test-suite.ts [hunk 25]
```diff
@@ -4,230 +4,227 @@ import { suite } from "../suite"
...
+        }
+        expect(mockDescribe).toHaveBeenCalledTimes(2)
+        mockDescribe.mock.calls[0][1]()
+        mockDescribe.mock.calls[1][1]()
+        expect(mockBeforeAll).toHaveBeenCalledTimes(3)
+        mockBeforeAll.mock.calls[0][0]()
+        mockBeforeAll.mock.calls[1][0]()
+        mockBeforeAll.mock.calls[2][0]()
+        // The beforeAll in SuperA is called twice in SuperA
+        // and A and the beforeAll in A is called only in A .
```
Comment (Prior99) [inline src/__tests__/test-suite.ts]: There is a space before the `.` here and it'd be cool if you'd wrap code-words in markdown style backticks:

```
// The `beforeAll` in `SuperA` is called twice in `SuperA`
// and `A` and the `beforeAll` in `A` is called only in `A`.
// Thus there should be three `beforeAll` calls.
```

File: src/test.ts [hunk 26]
```diff
@@ -5,10 +5,10 @@ import { configuration } from "./configure"
 export type TestFunction = (instance: any) => void
 
 export function getTests(target: any): TestFunction[] {
-    const tests = Reflect.getMetadata(`test-decorator:tests`, target)
+    const tests = Reflect.getMetadata(`test-decorators:tests`, target)
     if (typeof tests !== "undefined") { return tests }
     const newTests: TestFunction[] = []
-    Reflect.defineMetadata(`test-decorator:tests`, newTests, target)
+    Reflect.defineMetadata(`test-decorators:tests`, newTests, target)
```
Comment (Prior99) [inline src/test.ts]: Thank you :)

File: src/configure.ts [hunk 27]
```diff
@@ -14,11 +14,11 @@ export type DescribeFunction = () => Promise<void> | void
  */
 export interface ConfigureOptions {
     /**
-     * A function called before all tests
+     * The function called once per suite before all tests.
      */
     beforeAll?: BeforeAll,
     /**
-     * A function called for each test
+     * The function called once per suite before each test.
```
Comment (Prior99) [inline src/configure.ts]: No, it's called once per test, not once per suite.

**Review by Prior99 — CHANGES_REQUESTED** [review]

File: src/test.ts (hunk 26 above)
Comment (SergejKasper) [reply src/test.ts]: you're welcome :)

**Review by Prior99 — APPROVED** [review]
